src/oc/web/actions/user.cljs

```diff
@@ -60,4 +72,91 @@
     (dis/dispatch! [:login-with-slack])))
 
 (defn show-login [login-type]
-  (dis/dispatch! [:login-overlay-show login-type]))
\ No newline at end of file
+  (dis/dispatch! [:login-overlay-show login-type]))
+
+;; Auth
+
+(defn bot-auth []
+  (dis/dispatch! [:bot-auth]))
+
+(defn auth-with-token-failed [error]
+  (dis/dispatch! [:auth-with-token/failed error]))
+
+(defn auth-with-token-success [jwt]
+  (api/get-entry-point)
+  (api/get-auth-settings)
+  (when (= (:auth-with-token-type @dis/app-state) :password-reset)
+    (cook/set-cookie! :show-login-overlay "collect-password"))
+  (dis/dispatch! [:auth-with-token/success jwt]))
+
+(defn auth-with-token-callback
+  [success status body]
+  (if success
+    (do
+      (update-jwt body)
+      (auth-with-token-success body))
+    (cond
+      (= status 401)
+      (auth-with-token-failed 401)
+      :else
+      (auth-with-token-failed 500))))
+
+(defn auth-with-token [token-type]
+  (api/auth-with-token (:token (:query-params @router/path)) auth-with-token-callback)
+  (dis/dispatch! [:auth-with-token token-type]))
+
+;; Signup
+
+(defn signup-with-email-failed [status]
+  (dis/dispatch! [:signup-with-email/failed status]))
+
+(defn signup-with-email-success
+  [status jwt]
+  (cond
+    (= status 204) ;; Email wall since it's a valid signup w/ non verified email address
+    (utils/after 10 #(router/nav! (str oc-urls/email-wall "?e=" (:email (:signup-with-email @dis/app-state)))))
+    (= status 200) ;; Valid login, not signup, redirect to home
+    (if (or
+          (and (empty? (:first-name jwt)) (empty? (:last-name jwt)))
+          (empty? (:avatar-url jwt)))
+      (do
+        (utils/after 200 #(router/nav! oc-urls/sign-up-profile))
+        (api/get-entry-point))
+      (do
+        (api/get-entry-point)
+        (dis/dispatch! [:input [:email-lander-check-team-redirect] true])))
+    :else ;; Valid signup let's collect user data
+    (do
+      (cook/set-cookie! :jwt jwt (* 60 60 24 60) "/" ls/jwt-cookie-domain ls/jwt-cookie-secure)
+      (cook/set-cookie!
+       (router/show-nux-cookie (jwt/user-id))
+       (:new-user router/nux-cookie-values)
+       (* 60 60 24 7))
+      (utils/after 200 #(router/nav! oc-urls/sign-up-profile))
+      (api/get-entry-point)
+      (dis/dispatch! [:signup-with-email/success]))))
+
+(defn signup-with-email-callback
+  [success body status]
+  (if success
+    (signup-with-email-success status body)
+    (signup-with-email-failed status)))
+
+(defn signup-with-email []
+  (api/signup-with-email
+   (or (:firstname (:signup-with-email @dis/app-state)) "")
+   (or (:lastname (:signup-with-email @dis/app-state)) "")
+   (:email (:signup-with-email @dis/app-state))
+   (:pswd (:signup-with-email @dis/app-state))
+   signup-with-email-callback)
+  (dis/dispatch! [:signup-with-email]))
+
+(defn signup-with-email-data [signup-with-email-data]
```

Comment:
I think we can clean this up.  The above function `signup-with-email` should accept `firstname` `lastname` `email` `pswd` as parameters.  We can get rid of the `signup-with-email-data` function.  The signup component can maintain the data instead of having it in the app state. On successful signup we can pass the user data to the data store if we still need to keep this data.

Reply:
@thepug this is not possible because `signup-with-email` contains also the error states returned by the server. I agree this could be simplified but at this time it's beyond the scope of this refactor.
I changed the parameter name though because it was the same of the function name and could lead to problems.

Reply:
Are the error states only from the server?  You can still use the error states but no need to have the other bits in the app state, `firstname`, etc.